cpp/src/arrow/compute/kernels/scalar_cast_nested.cc

```diff
@@ -150,6 +150,104 @@ void AddListCast(CastFunction* func) {
+    const CastOptions& options = CastState::Get(ctx);
+    const auto in_field_count =
+        checked_cast<const StructType&>(*batch[0].type()).num_fields();
+    const auto out_field_count =
+        checked_cast<const StructType&>(*out->type()).num_fields();
+
+    if (in_field_count != out_field_count) {
+      return Status::TypeError("struct field sizes do not match: ",
+                               batch[0].type()->ToString(), " ", out->type()->ToString());
+    }
+
+    for (int i = 0; i < in_field_count; ++i) {
+      const auto in_field_name =
+          checked_cast<const StructType&>(*batch[0].type()).field(i)->name();
+      const auto out_field_name =
+          checked_cast<const StructType&>(*out->type()).field(i)->name();
+      if (in_field_name != out_field_name) {
+        return Status::TypeError(
+            "struct field names do not match: ", batch[0].type()->ToString(), " ",
+            out->type()->ToString());
+      }
+
+      const auto in_field_nullable =
+          checked_cast<const StructType&>(*batch[0].type()).field(i)->nullable();
+      const auto out_field_nullable =
+          checked_cast<const StructType&>(*out->type()).field(i)->nullable();
+
+      if (in_field_nullable && !out_field_nullable) {
+        return Status::TypeError("cannot cast nullable struct to non-nullable struct: ",
+                                 batch[0].type()->ToString(), " ",
+                                 out->type()->ToString());
+      }
+    }
+
+    for (int i = 0; i < in_field_count; ++i) {
+      const auto in_field_name =
+          checked_cast<const StructType&>(*batch[0].type()).field(i)->name();
+      const auto out_field_name =
+          checked_cast<const StructType&>(*out->type()).field(i)->name();
+      if (in_field_name != out_field_name) {
+        return Status::TypeError(
+            "struct field names do not match: ", batch[0].type()->ToString(), " ",
+            out->type()->ToString());
+      }
+    }
+
+    if (out->kind() == Datum::SCALAR) {
+      const auto& in_scalar = checked_cast<const StructScalar&>(*batch[0].scalar());
+      auto out_scalar = checked_cast<StructScalar*>(out->scalar().get());
+
+      DCHECK(!out_scalar->is_valid);
+      if (in_scalar.is_valid) {
+        for (int i = 0; i < in_field_count; i++) {
+          auto values = in_scalar.value[i];
+          auto target_type = out->type()->field(i)->type();
+          ARROW_ASSIGN_OR_RAISE(Datum cast_values,
+                                Cast(values, target_type, options, ctx->exec_context()));
+          DCHECK_EQ(Datum::SCALAR, cast_values.kind());
+          out_scalar->value.push_back(cast_values.scalar());
+        }
+        out_scalar->is_valid = true;
+      }
+      return Status::OK();
+    }
+
+    const ArrayData& in_array = *batch[0].array();
+    ArrayData* out_array = out->mutable_array();
+    out_array->buffers = in_array.buffers;
```

Comment:
Now that I look at this, I think there's still a missing case here. We should copy the offset over as well (and then we should not slice the child arrays below), or if there's a top-level validity buffer, we should slice the buffer before copying. I don't think the tests notice this right now because there's no top-level validity buffer in any of the test cases.

Reply:
`StructArray::Make` does take a null bitmap so we should be able to test this.

Reply:
First part of this is done. For the sake of comprehensiveness I tried introducing `null` values into the main test, but didn't make a difference at all. Is that different from manually providing a bitmap as you've described?

Reply:
It's different, yes, since StructArray is nested: for a given row, either the subarrays can be null, or the top-level StructArray itself may be null (and then the child array's nullity are ignored).

Reply:
Great explanation - think I've added this to the test now but didn't change the pass/fail status. Still missing something?

Reply:
This looks good now, since we're not slicing the child and copying over the out offset, so it should pass.

Reply:
Though - sorry - but it's probably better to slice the children, not copy the offset, and slice the bitmap in `in_array.buffers` when copying? The reason being, if we `Cast` a large array that's been sliced, this right now will cast the _entire_ child array, even though we only technically care about the sliced part. Does that make sense?

The null bitmap can be copied with CopyBitmap: https://github.com/apache/arrow/blob/ec38aebb36e99e54e69089cbc6a623a616575dde/cpp/src/arrow/util/bitmap_ops.h#L44-L46

It would be something like
```
if (in_array.buffers[0]) {
  ARROW_ASSIGN_OR_RAISE(out_array->buffers[0], CopyBitmap(ctx->memory_pool(), in_array.buffers[0]->data(), in_array.offset, in_array.length));
}
```

Though, frankly, it _should_ work below to set `kernel.null_handling` to `INTERSECTION` and then the compute infra will compute the null bitmap for you. (And you shouldn't have to mess with `out_array->buffers` or `out_array->offset`.)

Reply:
Sorry, there's lots of knobs to turn for this, so there's not one obvious way to implement things. I would see if `kernel.null_handling` + slicing children before casting works (it should, as far as I can tell without building and trying it myself.)

Reply:
No worries - this has been a great learning experience for me. Your code suggestion looks pretty similar to what's already being done with the List casts - I just didn't understand enough before why to include but makes sense here. Let me see how this all looks